### ISSUE TITLE
Compiler bug in release mode

### ISSUE THREAD

**Issue by CarnifexOptimus** [dotnet/runtime#109606]
### Description

The following method starts malfunctioning in release mode. It stops producing a correct circle arc, leading to all kind of errors in my code such as cones not drawing correctly, -1 cannot be greater than 0 and empty sequence errors.

### Reproduction Steps

the problematic method:    

public static WDir[] CircleArc(float radius, Angle angleStart, Angle angleEnd, float maxError)
    {
        var length = angleEnd - angleStart;
        var numSegments = CalculateCircleSegments(radius, length.Abs(), maxError);
        var angleIncrement = length / numSegments;
        var points = new WDir[numSegments + 1];
        for (var i = 0; i <= numSegments; ++i)
        {
            var angle = angleStart + i * angleIncrement;
            points[i] = PolarToCartesian(radius, angle);
        }
        return points;
    }

the full code can be found here: https://github.com/FFXIV-CombatReborn/BossmodReborn/blob/main/BossMod/Util/CurveApprox.cs

### Expected behavior

It should be generating a correct CircleArc as the method name suggests.

### Actual behavior

Results in incorrect circle arcs.

### Known Workarounds

If I add     [MethodImpl(MethodImplOptions.NoOptimization)] infront it or use debug mode it works correctly.
The code also works correctly in release mode if I add a debug log entry between var length and var numSegments or add a random useless if check such as if (angleStart == angleEnd) return [];

### Configuration

.Net version: 8.0.403
Windows 10 22H2 19045.5073
x64
it does not seem to be specific to my configuration, the bug also happens if I use a github workflow to build the project

**Comment by karakasa** [comment]
What are your parameters?

I tried `CurveApprox.CircleArc(10.0f, new Angle(0.0f), new Angle((float)Math.PI), 0.01f)` in Debug & Release mode but the results are the same.

**Comment by AndyAyersMS** [comment]
@CarnifexOptimus can you check which version of the runtime you're using? If you run `dotnet --info` look for something like:
```
  Microsoft.NETCore.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

**Comment by karakasa** [comment]
> [@CarnifexOptimus](https://github.com/CarnifexOptimus) can you check which version of the runtime you're using? If you run `dotnet --info` look for something like:
> 
> ```
>   Microsoft.NETCore.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
> ```

isn’t it 8.0.10? The original post writes ‘8.0.403’

**Comment by am11** [comment]
`8.0.403` is the SDK version; runtime versions are different.

**Comment by CarnifexOptimus** [comment]
@karakasa Thank you for taking the time to test. The bug seems to happen with any parameters, some are hardcoded, some are dynamically read from memory. The bug is reproduceable 100% on my machine. If I remove the [MethodImpl(MethodImplOptions.NoOptimization)] line, all instance where CircleArc is used in the project results in bugs.

Here is an example of the parameters used. In this case all parameters are hardcoded:
https://github.com/FFXIV-CombatReborn/BossmodReborn/blob/main/BossMod/Modules/Shadowbringers/Foray/DelubrumReginaeSavage/DRS7StygimolochLord/Border.cs

Here the bug results in an empty polygon and -1 cannot be greater than 0 errors. In other places it may only lead to less severe errors such as the cones being drawn in incorrect and seemingly random sizes. I uploaded an example. Left is bugged and right correct size without optimisations. As you can see on the left side the cones are much smaller and seemingly have a random radius, while having a hardcoded radius of 18 and 60° halfangle.
The rotation offset from source and world position are read from memory.
![Image](https://github.com/user-attachments/assets/5363fc46-09d8-459f-a997-9a98a5b84456)



@AndyAyersMS according to dotnet --info the runtime version is 8.0.10

**Comment by janvorli** [comment]
Maybe the reason why @karakasa couldn't repro it might be the AVX level support in the processor. Maybe AVX512 codegen has a problem and regular AVX doesn't or vice versa. Just guessing.

**Comment by CarnifexOptimus** [comment]
My CPU does not support AVX512, it is an Intel i7-8700k. That should (hopefully?) rule out a problem with newer instruction sets.

**Comment by AndyAyersMS** [comment]
FYI @dotnet/jit-contrib

**Comment by kunalspathak** [comment]
> I tried `CurveApprox.CircleArc(10.0f, new Angle(0.0f), new Angle((float)Math.PI), 0.01f)` in Debug & Release mode and the results are the same.

same for me. Do you mind creating a shorter repro and send debug vs. release output?

**Comment by CarnifexOptimus** [comment]
@kunalspathak I can try to make a shorter repro, but the problem seems to be that even small changes to the code make the bug disappear. If I use an IEnumerable instead of an array the bug no longer happens. If I add additional debug steps to the code (that simply print the result of each step to a log file) the bug no longer occurs. If I add redundant if checks between var length and var numSegments such as if (angleStart == angleEnd) return []; the bug no longer happens. To me it seems like it is optimising a crucial part of the code away under a very specific condition which I unluckily encountered here. It is the first time since I started coding that I noticed such a compiler bug, so I am not sure how to proceed from here.
I will be honest and say that I know very little about how the JIT compiler works, so I cannot say how seemingly not directly connected code in the project might affect the output after all the optimisations. Considering several people already tested a cut down version of the code it suggests this problem is caused in connection with other code only.

I made a test class inside my project with the exact values the users tested here (CurveApprox.CircleArc(10.0f, new Angle(0.0f), new Angle((float)Math.PI), 0.01f) and it results in a cone that got a radius of about 3.33 instead of 10 if I do not disabled the optimisations:

![Image](https://github.com/user-attachments/assets/7fb60b9a-16ca-479f-9328-11b54dddf221)

so as seen in my previous example screenshot it seems like it is applying a random scale factor to the arc somehow.

**Comment by CarnifexOptimus** [comment]
Here are the vertices (CurveApprox.CircleArc(10.0f, new Angle(0.0f), new Angle((float)Math.PI), 0.01f) outputs for me rounded to 3 decimals:

with disabled optimisations it outputs the following vertices:

(0,000, 10,000), (0,872, 9,962), (1,736, 9,848), (2,588, 9,659), (3,420, 9,397), (4,226, 9,063), (5,000, 8,660), (5,736, 8,192), (6,428, 7,660), (7,071, 7,071), (7,660, 6,428), (8,192, 5,736), (8,660, 5,000), (9,063, 4,226), (9,397, 3,420), (9,659, 2,588), (9,848, 1,736), (9,962, 0,872), (10,000, -0,000), (9,962, -0,872), (9,848, -1,736), (9,659, -2,588), (9,397, -3,420), (9,063, -4,226), (8,660, -5,000), (8,192, -5,736), (7,660, -6,428), (7,071, -7,071), (6,428, -7,660), (5,736, -8,192), (5,000, -8,660), (4,226, -9,063), (3,420, -9,397), (2,588, -9,659), (1,736, -9,848), (0,872, -9,962), (-0,000, -10,000)

with enabled optimisations the output is:
(0,000, 3,142), (0,491, 3,103), (0,971, 2,988), (1,426, 2,799), (1,847, 2,542), (2,221, 2,221), (2,542, 1,847), (2,799, 1,426), (2,988, 0,971), (3,103, 0,491), (3,142, -0,000), (3,103, -0,491), (2,988, -0,971), (2,799, -1,426), (2,542, -1,847), (2,221, -2,221), (1,847, -2,542), (1,426, -2,799), (0,971, -2,988), (0,491, -3,103), (-0,000, -3,142)

my conclusion is that it it inputting incorrect values (likely the radius given the visual output) into var numSegments = CalculateCircleSegments(radius, length.Abs(), maxError); since the amount of vertices is based on radius and the full angle of the circle arc.

**Comment by CarnifexOptimus** [comment]
with  (CurveApprox.CircleArc(20.0f, new Angle(0.0f), new Angle((float)Math.PI), 0.01f) it still outputs pi as the radius:

(0,000, 3,142), (0,491, 3,103), (0,971, 2,988), (1,426, 2,799), (1,847, 2,542), (2,221, 2,221), (2,542, 1,847), (2,799, 1,426), (2,988, 0,971), (3,103, 0,491), (3,142, -0,000), (3,103, -0,491), (2,988, -0,971), (2,799, -1,426), (2,542, -1,847), (2,221, -2,221), (1,847, -2,542), (1,426, -2,799), (0,971, -2,988), (0,491, -3,103), (-0,000, -3,142)

for (CurveApprox.CircleArc(10.0f, 0.Degrees(), 45.Degrees(), 0.01f) with optimisations it outputs:
(0,000, 0,785), (0,153, 0,770), (0,301, 0,726), (0,436, 0,653), (0,555, 0,555)

without optimisations:
(0,000, 10,000), (0,785, 9,969), (1,564, 9,877), (2,334, 9,724), (3,090, 9,511), (3,827, 9,239), (4,540, 8,910), (5,225, 8,526), (5,878, 8,090), (6,494, 7,604), (7,071, 7,071)

now the radius is pi/4

for (CurveApprox.CircleArc(10.0f, 0.Degrees(), 360.Degrees(), 0.01f) without optimisations it outputs:
(0,000, 10,000), (0,872, 9,962), (1,736, 9,848), (2,588, 9,659), (3,420, 9,397), (4,226, 9,063), (5,000, 8,660), (5,736, 8,192), (6,428, 7,660), (7,071, 7,071), (7,660, 6,428), (8,192, 5,736), (8,660, 5,000), (9,063, 4,226), (9,397, 3,420), (9,659, 2,588), (9,848, 1,736), (9,962, 0,872), (10,000, -0,000), (9,962, -0,872), (9,848, -1,736), (9,659, -2,588), (9,397, -3,420), (9,063, -4,226), (8,660, -5,000), (8,192, -5,736), (7,660, -6,428), (7,071, -7,071), (6,428, -7,660), (5,736, -8,192), (5,000, -8,660), (4,226, -9,063), (3,420, -9,397), (2,588, -9,659), (1,736, -9,848), (0,872, -9,962), (-0,000, -10,000), (-0,872, -9,962), (-1,736, -9,848), (-2,588, -9,659), (-3,420, -9,397), (-4,226, -9,063), (-5,000, -8,660), (-5,736, -8,192), (-6,428, -7,660), (-7,071, -7,071), (-7,660, -6,428), (-8,192, -5,736), (-8,660, -5,000), (-9,063, -4,226), (-9,397, -3,420), (-9,659, -2,588), (-9,848, -1,736), (-9,962, -0,872), (-10,000, 0,000), (-9,962, 0,872), (-9,848, 1,736), (-9,659, 2,588), (-9,397, 3,420), (-9,063, 4,226), (-8,660, 5,000), (-8,192, 5,736), (-7,660, 6,428), (-7,071, 7,071), (-6,428, 7,660), (-5,736, 8,192), (-5,000, 8,660), (-4,226, 9,063), (-3,420, 9,397), (-2,588, 9,659), (-1,736, 9,848), (-0,872, 9,962), (0,000, 10,000)

with optimisations:
(0,000, 6,283), (0,703, 6,244), (1,398, 6,126), (2,075, 5,931), (2,726, 5,661), (3,343, 5,320), (3,918, 4,912), (4,443, 4,443), (4,912, 3,918), (5,320, 3,343), (5,661, 2,726), (5,931, 2,075), (6,126, 1,398), (6,244, 0,703), (6,283, -0,000), (6,244, -0,703), (6,126, -1,398), (5,931, -2,075), (5,661, -2,726), (5,320, -3,343), (4,912, -3,918), (4,443, -4,443), (3,918, -4,912), (3,343, -5,320), (2,726, -5,661), (2,075, -5,931), (1,398, -6,126), (0,703, -6,244), (-0,000, -6,283), (-0,703, -6,244), (-1,398, -6,126), (-2,075, -5,931), (-2,726, -5,661), (-3,343, -5,320), (-3,918, -4,912), (-4,443, -4,443), (-4,912, -3,918), (-5,320, -3,343), (-5,661, -2,726), (-5,931, -2,075), (-6,126, -1,398), (-6,244, -0,703), (-6,283, 0,000), (-6,244, 0,703), (-6,126, 1,398), (-5,931, 2,075), (-5,661, 2,726), (-5,320, 3,343), (-4,912, 3,918), (-4,443, 4,443), (-3,918, 4,912), (-3,343, 5,320), (-2,726, 5,661), (-2,075, 5,931), (-1,398, 6,126), (-0,703, 6,244), (0,000, 6,283)

now 6.283 is 2* pi, so twice as much as for the 180° arc.

Conclusion: For whatever reason it inputs var length instead of radius into var numSegments = CalculateCircleSegments(radius, length.Abs(), maxError) when optimisations are enabled. Definitely one of the oddest bugs I have ever seen.

**Comment by karakasa** [comment]
> Conclusion: It likely inputs angleEnd instead of the radius into var numSegments = CalculateCircleSegments(radius, length.Abs(), maxError) when optimisations are enabled. Definitely one of the oddest bugs I have ever seen.

You can write a simple call to `CircleArc()` and use `DisassemblyDiagnoser` from `BenchmarkDotNet` ([guide](https://adamsitnik.com/Disassembly-Diagnoser/)) to get the output assembly and post them here.

**Comment by janvorli** [comment]
@CarnifexOptimus I have just noticed that your repo already contains a commit that works around the issue using the ` [MethodImpl(MethodImplOptions.NoOptimization)]`. @kunalspathak, @karakasa I wonder if you could have attempted to repro it with that state of the repo, as it would obviously not repro in that case.

**Comment by karakasa** [comment]
> `[MethodImpl(MethodImplOptions.NoOptimization)]`. [@kunalspathak](https://github.com/kunalspathak), [@karakasa](https://github.com/karakasa) I wonder if you could have attempted to repro it with that state of the repo, as it would obviously not repro in that case.

I tried Debug & Release & NoOptimization+Release & AggressiveOptimization+Release while four have the same output.

**Comment by janvorli** [comment]
@CarnifexOptimus how do I successfully build your repo? I've tried to run dotnet build in the root directory, but I get a lot of build errors, for example "The type or namespace name 'Lumina' could not be found " and many others.

**Comment by CarnifexOptimus** [comment]
@janvorli I believe you need to download https://goatcorp.github.io/dalamud-distrib/latest.zip and extract it to AppData\XIVLauncher\addon\Hooks\dev as seen in the workflow file: https://github.com/FFXIV-CombatReborn/BossmodReborn/blob/main/.github/workflows/publish.yml
Not sure why it was made this way. We only forked someone else's project back in April and never seen a reason to rework it.

**Comment by janvorli** [comment]
@CarnifexOptimus thank you, that helped to build it. So how do I run it to repro the problem?

**Comment by CarnifexOptimus** [comment]
@karakasa I tried to make a shorter repro, but as suspected the bug does not seem to trigger in different code. I also tested your minimal code from your initial post and it works just fine for both release and debug.
I wonder if it is even worth pursuing this problem further if it only happens under such nebulous and specific conditions.

@janvorli that will be tricky unless you already happen to have final fantasy 14 installed. You could create a free to play account, then install the game and XIVlauncher. Then use this 3rd party launcher to run the game. When running the game with XIVlauncher you can load plugins like this. Theoretically the UIDev part of the project lets you test various functionality of the plugin without logging into the game with a 3rd party launcher. But I already tried to repro this bug in UIDev and it does not seem to trigger under those circumstances. This further underlines that the JIT compiler is doing a really whacky optimisation there that only happens under very specific conditions.

I already had other users test a version of this project with optimisations enabled and they see the same bug, so it is unlikely to be triggered by specific hardware or net versions.

**Comment by AndyAyersMS** [comment]
@CarnifexOptimus can you try an experiment, since we are not yet able to reproduce the issue?

First, make a temp directory somewhere (eg `c:\temp`).

First, for the version that has the problem, set `DOTNET_JitDisasm=CircleArc` and `DOTNET_JitStdOutFile=c:\temp\bad.dasm` in the environment where `dotnet` is executed. Then run and verify things fail.

Second, for a version with minor changes that doesn't fail, set `DOTNET_JitDisasm=CircleArc` and `DOTNET_JitStdOutFile=c:\temp\good.dasm`  and run and verify things succeed.

Then share `bad.dasm` and `good.dasm` so we can see what is different; from there perhaps we can get an idea why.

**Comment by CarnifexOptimus** [comment]
@AndyAyersMS I created 3 versions:

release, bugged
[bad.txt](https://github.com/user-attachments/files/17688348/bad.txt)
release with the var radius = Radius workaround, works
[good.txt](https://github.com/user-attachments/files/17688349/good.txt)
debug, works
[goodDebug.txt](https://github.com/user-attachments/files/17688350/goodDebug.txt)

renamed the .dasm files to .txt since it doesn't allow to upload dasm files here.

I hope these help.

**Comment by AndyAyersMS** [comment]
@CarnifexOptimus thanks. In the "bad" codegen, the input register `xmm0` is modified in the prolog before it is used (diff below is good on left, bad on right). That likely explains why you are seeing the wrong results.

![Image](https://github.com/user-attachments/assets/569822d1-92ea-4f8a-a78b-b097db6a943d)

However when I try and repro locally, so far, I can only ever get a good prolog sequence (though different from either of yours).
```asm
;; bad
       vmovd    xmm7, edx
       vmovd    xmm0, r8d
       vmovaps  xmm6, xmm0

;; good
       vmovd    xmm6, edx
       vmovd    xmm1, r8d

;; mine
       vmovaps  xmm6, xmm0
       vmovd    xmm7, edx
       vmovd    xmm0, r8d
```
This seems similar to what we fixed in #98749, which I believe first appeared in 8.0.5.

Do you have multiple 8.0 runtimes installed on the box? If so, can you double-check that you are not somehow running against an older 8.0 release?

Does the problem go away (with the "bad" version) if you run with `DOTNET_JitEnablePhysicalPromotion=0`?

**Comment by CarnifexOptimus** [comment]
> [@CarnifexOptimus](https://github.com/CarnifexOptimus) I see you closed this -- is the issue resolved or ...?

Thank you for asking. After you mentioning that the bug might have been fixed in 8.0.5 I did some digging and found out XIVLauncher is loading a 8.0.0 runtime internally despite having my computer having 8.0.10 installed. This explains why noone could not reproduce it outside of this project. I informed the XIVLauncher maintainers about this bug in 8.0.0 and they said they plan to upgrade to Net9 soon, so the problem should hopefully be fully resolved then.
